network/src/connection_manager.rs

```diff
@@ -0,0 +1,404 @@
+        disconnected_peers
+    }
+
+    /// TODO (howardwu): Add logic to remove the active channels
+    ///  and handshakes of the peer from this struct.
+    /// Disconnects the given address from the node.
+    #[inline]
+    pub async fn disconnect_from_peer(&self, remote_address: &SocketAddr) -> bool {
+        // Acquire the peer book write lock.
+        let mut peer_book = self.peer_book.write().await;
+        // Disconnected from the peer.
+        let is_disconnected = peer_book.disconnected_peer(remote_address);
+        // Drop the peer book write lock.
+        drop(peer_book);
+        is_disconnected
+    }
+
+    /// Attempts to get the channel for a given address.
+    /// Returns `Some(channel)` if the address is a connected peer.
+    /// Otherwise, returns `None`.
+    #[inline]
+    pub async fn get_channel(&self, remote_address: &SocketAddr) -> Option<Arc<Channel>> {
+        // TODO (howardwu): Remove this logic in favor of the below calling convention.
+        //  This is a temporary solution as part of a much larger refactor.
+        // self.channels.get(remote_address)
+
+        // Acquire the tmp_connections read lock.
+        let tmp_connections = self.tmp_connections.read().await;
+        // Forget the local address.
+        let channel = tmp_connections.get(remote_address);
+        // Drop the tmp_connections read lock.
+        drop(tmp_connections);
+        channel
+    }
+
+    /// Stores a new channel at the peer address it is connected to.
+    pub fn add_channel(&mut self, channel: &Arc<Channel>) {
+        self.channels.insert(channel.address, channel.clone());
+    }
+
+    /// Manages all peer connections and processes updates with all connected peers.
+    #[inline]
+    pub async fn handler(&self) {
+        // If the node is connected to less peers than the minimum required,
+        // ask every peer the node is connected to for more peers.
+        if self.get_num_connected().await < self.minimum_peer_count {
+            // Broadcast a `GetPeers` message to request for more peers.
+            self.broadcast_getpeers_requests().await;
+            // Attempt a connection request with every disconnected peer.
+            self.connect_to_all_disconnected_peers().await;
+            // Attempt a connection request with every bootnode peer again.
+            // The goal is to reconnect with any bootnode peers we might
+            // have failed to connect to. The manager will filter attempts
+            // to connect with itself or already connected bootnode peers.
+            self.connect_to_bootnodes().await;
+        }
+
+        // TODO (howardwu): Unify `Ping` and `Version`requests.
+        //  This is a remnant and these currently do not need to be distinct.
+
+        // Broadcast a `Ping` request to each connected peer.
+        self.broadcast_ping_requests().await;
+        // Broadcast a `Version` request to each connected peer.
+        self.broadcast_version_requests().await;
```

Comment:
some of these methods can probably be run concurrently (as opposed to halting the current task with subsequent `await`s) with `tokio::join`; in addition, if any of them are expected to be executed in tandem, like e.g. `broadcast_ping_requests` and `broadcast_version_requests`, they could reuse some of the results of their internal calls, e.g. to `get_all_connected`; this could be achieved with merging them into a single method or by passing the common, previously obtained values to both